README.md
** Implement Guide **

* Step 1: Start zookeeper, kafka-server
- bin/zookeeper-server-start.sh config/zookeeper.properties
- bin/kafka-server-start.sh config/server.properties

* Step 2: Create topic
- bin/kafka-topics.sh --create --topic result --bootstrap-server localhost:9092

* Step 3: Start questdb and questdb-kafka-connector
- bin/connect-standalone.sh config/connect-standalone.properties config/questdb-connector.properties

* Step 4: Run producer and streaming by spark-submit
- python3 producer.py
- spark-submit --packages org.apache.spark:spark-sql-kafka-0-10_2.12:3.3.1 stream.py
  
* Step 5: Visualize data from questdb
- run getData.ipynb
- bin/kafka-console-consumer.sh --topic result --bootstrap-server localhost:9092


conda-meta/history
==> 2023-02-20 20:50:47 <==
# cmd: /usr/bin/conda create -p /home/bakansm/Code/kafka-ml-test
# conda version: 4.14.0


test.py
from pyspark.sql import Window
from pyspark.sql.types import *
from pyspark.sql.functions import *
from pyspark.sql import SparkSession
from config import config, params
from pyspark.ml.feature import StringIndexer
from pyspark.ml.feature import VectorAssembler


KAFKA_BOOTSTRAP_SERVER = "localhost:9092"
KAFKA_TOPIC = "{0},{1},{2},{3},{4},{5},{6}" \
              .format(config['topic_1'],
                      config['topic_2'],
                      config['topic_3'],
                      config['topic_4'],
                      config['topic_5'],
                      config['topic_6'],
                      config['topic_7'])
CHECKPOINT_LOCATION = "./checkpoint"

### Khởi tạo spark
spark = SparkSession \
    .builder \
    .appName("Kafka streaming test") \
    .master('local[*]') \
    .getOrCreate()


### Tạo khung dataframe
schema = StructType([ 
  StructField("timestamp", TimestampType(), True),
  StructField("currency" , StringType(), True),
  StructField("amount" , FloatType(), True),
])


### Lấy dữ liệu từ kafka ( định dạng json )
jsonData = spark \
  .readStream \
  .format("kafka") \
  .option("kafka.bootstrap.servers", KAFKA_BOOTSTRAP_SERVER) \
  .option("subscribe", KAFKA_TOPIC) \
  .option("startingOffsets", "earliest") \
  .option("includeHeaders", "true") \
  .option("failOnDataLoss","false") \
  .load()

### Chuyển json sang datafram pyspark
dataframe = jsonData.select(from_json(col("value").cast("string"), schema).alias("value")).select(col("value.*"))

# dataframe = dataframe.groupBy('currency').agg(avg('amount').cast('Float').alias('average'), current_timestamp().alias('date'))

# dataframe.printSchema()

# ############################# IMPORTANT #############################
# ### ---> Code mô hình máy học ở đây khúc này bằng dataframe  <--- ###
# #####################################################################

### Chuyển dataframe ngược lại json để thêm vào topic
reward_data = dataframe.select(to_json(struct("*")).alias("value"))

### Thêm data vào lại topic
ds = reward_data \
  .writeStream \
  .format("kafka") \
  .trigger(processingTime="1 seconds") \
  .option("kafka.bootstrap.servers", KAFKA_BOOTSTRAP_SERVER) \
  .option("checkpointLocation", CHECKPOINT_LOCATION) \
  .outputMode("update") \
  .option("topic", "result") \
  .start()

ds.awaitTermination()

consumer.py
import pandas as pd 
import numpy as np 
import datetime as dt

from kafkaHelper import produceRecord, consumeRecord, initConsumer, initProducer
from config import config, params

# initialize Kafka consumers and producer
print('Starting Apache Kafka consumers and producer')
consumer_1 = initConsumer(config['topic_1'])
# consumer_2 = initConsumer(config['topic_2'])
# consumer_3 = initConsumer(config['topic_3'])
producer = initProducer()

# intialize local dataframe
data_1 = pd.DataFrame(columns=['time', 'value'])
# data_2 = pd.DataFrame(columns=['time', 'value'])
# data_3 = pd.DataFrame(columns=['time', 'value'])

while True:
    # consume data from Kafka
    # topic 1 --> 4
    records_1 = consumeRecord(consumer_1)
    print('Consume record from topic \'{0}\' at time {1}'.format(config['topic_1'], dt.datetime.utcnow()))
    for r in records_1:
        dt_obj = dt.datetime.fromtimestamp(r['timestamp'])
        data_1.loc[len(data_1)] = [int(dt_obj.timestamp()), float(r['amount'])]
        ma_1 = {'timestamp': r['timestamp'], 'amount': float(data_1['value'].tail(n=params['ma']).mean())}
        # produce data
        produceRecord(ma_1, producer, config['topic_4'])
        print('Produce record to topic \'{0}\' at time {1}'.format(config['topic_4'], dt.datetime.utcnow()))

    # # topic 2 --> 5
    # records_2 = consumeRecord(consumer_2)
    # print('Consume records from topic \'{0}\' at time {1}'.format(config['topic_2'], dt.datetime.utcnow()))
    # for r in records_2:
    #     dt_obj = dt.datetime.fromtimestamp(r['timestamp'])
    #     data_2.loc[len(data_2)] = [int(dt_obj.timestamp()), float(r['amount'])]
    #     ma_2 = {'timestamp': r['timestamp'], 'amount': float(data_2['value'].tail(n=params['ma']).mean())}
    #     # produce data
    #     produceRecord(ma_2, producer, config['topic_5'])
    #     print('Produce record to topic \'{0}\' at time {1}'.format(config['topic_5'], dt.datetime.utcnow()))

    # # topic 3 --> 6
    # records_3 = consumeRecord(consumer_3)
    # print('Consume records from topic \'{0}\' at time {1}'.format(config['topic_3'], dt.datetime.utcnow()))
    # for r in records_3:
    #     dt_obj = dt.datetime.fromtimestamp(r['timestamp'])
    #     data_3.loc[len(data_3)] = [int(dt_obj.timestamp()), float(r['amount'])]
    #     ma_3 = {'timestamp': r['timestamp'], 'amount': float(data_3['value'].tail(n=params['ma']).mean())}
    #     # produce data
    #     produceRecord(ma_3, producer, config['topic_6'])
    #     print('Produce record to topic \'{0}\' at time {1}'.format(config['topic_6'], dt.datetime.utcnow()))

producer.py
import json, requests, time, asyncio
import datetime as dt
import pandas as pd
import numpy as np
import os

from sklearn.preprocessing import MinMaxScaler
from tensorflow.keras.models import load_model
from kafkaHelper import initProducer, produceRecord
from config import config, params

os.system('clear')

# real time data collector
async def async_getCryptoRealTimeData(producer, topic, param, time_inverval):
    while True:
        t_0 = time.time()
        # call API
        uri = 'https://api.binance.com/api/v3/ticker'
        # print('URI: ' + uri )


        # Store 1 hour API data
        datetime=[]
        open=[]
        high=[]
        low=[]
        volume=[]
        vwap=[]

        for i in range(0,10):
            res = requests.get(uri,params=param)
            if (res.status_code==200):            # read json response
                # read json response
                data = json.loads(res.content)
                date = dt.datetime.fromtimestamp(data['openTime']/1000)
                # add schema
                datetime.append(str(date))
                open.append(float(data['openPrice']))
                high.append(float(data['highPrice']))
                low.append(float(data['lowPrice']))
                volume.append(float(data['volume']))
                vwap.append(float(data['weightedAvgPrice']))
                print('Get {0}/60 Data'.format(i+1))
                time.sleep(60)
            else:
                # debug / print message
                print('Failed API request at time {0}'.format(dt.datetime.utcnow()))
                print('======================================')
                time.sleep(60)
        df = pd.DataFrame({ 'Date': datetime,
                            'Open': open,
                            'High': high,
                            'Low': low,
                            'Volume BTC': volume,
                            'Vwap': vwap
                        })
        inputData = df.set_index("Date")


        # Create X, Y scaler
        scalerInput = MinMaxScaler(feature_range=(0,1))
        scalerOuput = MinMaxScaler(feature_range=(0,1))

        scalerOuput_features = inputData[['Open']]
        scalerOuput_features = scalerOuput_features.values
        scalerOuput_features = scalerOuput.fit_transform(scalerOuput_features)

        inputData = scalerInput.fit_transform(inputData)

        inputData=np.array(inputData)
        inputData=np.reshape(inputData,(inputData.shape[0],inputData.shape[1],1))
        model = load_model('btc-predict-model.h5')
        pred = model.predict(inputData)
        pred = scalerOuput.inverse_transform(pred)
        df['Predict'] = pred
        print(df)
        df.reset_index()
        for index, row in df.iterrows():
            message =   {   
                            'Date': row['Date'],
                            'Open': row['Open'],
                            'High': row['High'],
                            'Low': row['Low'],
                            'Volume BTC': row['Volume BTC'],
                            'Vwap': row['Vwap'],
                            'Predict': row['Predict']
                        }
            # produce record to kafka
            produceRecord(message, producer, topic)
        
        # wait
        await asyncio.sleep(time_inverval - (time.time() - t_0))

# initialize kafka producer
producer = initProducer()

# define async routine
async def main():
    await asyncio.gather(
    async_getCryptoRealTimeData(producer, config['topic_1'], params['param'], params['api_call_period'])
)

# run async routine
asyncio.run(main())

streamer.py
from pyspark.sql.types import *
from pyspark.sql.functions import *
from pyspark.sql import SparkSession
from config import config, params
from pyspark.sql.functions import col
from config import config, params

KAFKA_BOOTSTRAP_SERVER = "localhost:9092"

spark = SparkSession \
    .builder \
    .appName("Kafka streaming test") \
    .master('local[*]') \
    .getOrCreate()

reader = spark \
  .readStream \
  .format("kafka") \
  .option("header", "true") \
  .option("kafka.bootstrap.servers", KAFKA_BOOTSTRAP_SERVER) \
  .option("subscribe", config["topic_1"]) \
  .option("startingOffsets", "latest") \
  .load()

ds = reader \
  .selectExpr("CAST(key AS STRING)", "CAST(value AS STRING)") \
  .writeStream \
  .format("kafka") \
  .option("kafka.bootstrap.servers", "localhost:9092") \
  .option("checkpointLocation", "/home/bakansm/Documents/kafka-ml-test/checkpoint") \
  .option("topic", "topic_BTC") \
  .start()

ds.awaitTermination()

stream_BTC.py
from pyspark.sql.types import *
from pyspark.sql.functions import *
from pyspark.sql import SparkSession
from config import config, params
from pyspark.sql.functions import col
from config import config, params

KAFKA_BOOTSTRAP_SERVER = "localhost:9092"
INPUT_KAFKA_TOPIC = config["topic_1"]
OUTPUT_KAFKA_TOPIC = 'result_BTC'
CHECKPOINT_LOCATION = "./checkpoint"


### Khởi tạo spark
spark = SparkSession \
    .builder \
    .appName("Kafka streaming test") \
    .master('local[*]') \
    .getOrCreate()


### Tạo khung dataframe
schema = StructType([ 
  StructField("timestamp", StringType(), True),
  StructField("currency" , StringType(), True),
  StructField("amount" , StringType(), True),
])


### Lấy dữ liệu từ kafka ( định dạng json )
jsonData = spark \
  .readStream \
  .format("kafka") \
  .option("kafka.bootstrap.servers", KAFKA_BOOTSTRAP_SERVER) \
  .option("subscribe", INPUT_KAFKA_TOPIC) \
  .option("startingOffsets", "latest") \
  .option("includeHeaders", "true") \
  .option("failOnDataLoss","false") \
  .load()

### Chuyển json sang datafram pyspark
dataframe = jsonData.select(from_json(col("value").cast("string"), schema).alias("value"))




############################# IMPORTANT #############################
### ---> Code mô hình máy học ở đây khúc này bằng dataframe  <--- ###
#####################################################################

## Chuyển dataframe ngược lại json để thêm vào topic
forward_data = dataframe.select(to_json(struct("value.*")).alias("value"))

### Thêm data vào lại topic
ds = forward_data \
  .writeStream \
  .format("kafka") \
  .trigger(processingTime="5 seconds") \
  .option("kafka.bootstrap.servers", KAFKA_BOOTSTRAP_SERVER) \
  .option("checkpointLocation", CHECKPOINT_LOCATION) \
  .outputMode("append") \
  .option("topic", OUTPUT_KAFKA_TOPIC) \
  .start()

ds.awaitTermination()

stream_ETH.py
from pyspark.sql.types import *
from pyspark.sql.functions import *
from pyspark.sql import SparkSession
from config import config, params
from pyspark.sql.functions import col
from config import config, params

KAFKA_BOOTSTRAP_SERVER = "localhost:9092"
INPUT_KAFKA_TOPIC = config["topic_2"]
OUTPUT_KAFKA_TOPIC = 'result'
CHECKPOINT_LOCATION = "./checkpoint"


### Khởi tạo spark
spark = SparkSession \
    .builder \
    .appName("Kafka streaming test") \
    .master('local[*]') \
    .getOrCreate()


### Tạo khung dataframe
schema = StructType([ 
  StructField("timestamp", StringType(), True),
  StructField("currency" , StringType(), True),
  StructField("amount" , StringType(), True),
])


### Lấy dữ liệu từ kafka ( định dạng json )
jsonData = spark \
  .readStream \
  .format("kafka") \
  .option("kafka.bootstrap.servers", KAFKA_BOOTSTRAP_SERVER) \
  .option("subscribe", INPUT_KAFKA_TOPIC) \
  .option("startingOffsets", "latest") \
  .option("includeHeaders", "true") \
  .option("failOnDataLoss","false") \
  .load()

### Chuyển json sang datafram pyspark
dataframe = jsonData.select(from_json(col("value").cast("string"), schema).alias("value"))


############################# IMPORTANT #############################
### ---> Code mô hình máy học ở đây khúc này bằng dataframe  <--- ###
#####################################################################


### Chuyển dataframe ngược lại json để thêm vào topic
forward_data = dataframe.select(to_json(struct("value.*")).alias("value"))

### Thêm data vào lại topic
ds = forward_data \
  .writeStream \
  .format("kafka") \
  .trigger(processingTime="5 seconds") \
  .option("kafka.bootstrap.servers", KAFKA_BOOTSTRAP_SERVER) \
  .option("checkpointLocation", CHECKPOINT_LOCATION) \
  .outputMode("append") \
  .option("topic", OUTPUT_KAFKA_TOPIC) \
  .start()

ds.awaitTermination()

stream.py
from pyspark.sql.types import *
from pyspark.sql.functions import *
from pyspark.sql import SparkSession
from config import config, params
import pandas as pd
import numpy as np
from sklearn.preprocessing import MinMaxScaler

KAFKA_BOOTSTRAP_SERVER = "localhost:9092"
KAFKA_TOPIC = config['topic_1']
CHECKPOINT_LOCATION = "./checkpoint"

### Khởi tạo spark
spark = SparkSession \
    .builder \
    .appName("Kafka streaming test") \
    .master('local[*]') \
    .getOrCreate()


### Tạo khung dataframe
schema = StructType([ 
  StructField("Date", StringType(), True),
  StructField("Open", FloatType(), True),
  StructField("High", FloatType(), True),
  StructField("Low", FloatType(), True),
  StructField("Volume BTC", FloatType(), True),
  StructField("Vwap", FloatType(), True),
  StructField("Predict", FloatType(), True),
])


### Lấy dữ liệu từ kafka ( định dạng json )
data = spark \
  .readStream \
  .format("kafka") \
  .option("kafka.bootstrap.servers", KAFKA_BOOTSTRAP_SERVER) \
  .option("subscribe", KAFKA_TOPIC) \
  .option("startingOffsets", "earliest") \
  .option("includeHeaders", "true") \
  .option("failOnDataLoss","false") \
  .load()

### Chuyển json sang datafram pyspark
dataframe = data.select(from_json(col("value").cast("string"), schema).alias("value")).select(col("value.*"))

############################# IMPORTANT #############################
### ---> Code mô hình máy học ở đây khúc này bằng dataframe  <--- ###
#####################################################################

# ### Chuyển dataframe ngược lại json để thêm vào topic
reward_data = dataframe.select(to_json(struct("*")).alias("value"))

# ### Thêm data vào lại topic
ds = reward_data \
  .writeStream \
  .format("kafka") \
  .trigger(processingTime="60 seconds") \
  .option("kafka.bootstrap.servers", KAFKA_BOOTSTRAP_SERVER) \
  .option("checkpointLocation", CHECKPOINT_LOCATION) \
  .outputMode("update") \
  .option("topic", "result") \
  .start()

ds.awaitTermination()

# result_ds = result \
#   .writeStream \
#   .format(“csv”) \
#   .outputMode(“append”) \
#   .option(“path”, "/home/bakansm/Code/kafka-ml-test/result.csv") \
#   .option(“checkpointLocation”, CHECKPOINT_LOCATION) \
#   .start()